### PR TITLE
Azure OpenAI: update codegen

### PR DESCRIPTION
**Pull request by trrwilson**
This change updates the typespec basis for Azure OpenAI to the latest from `azure-rest-api-specs`. That new version includes proper definition of models for Chat API objects as well as better-integrated rename, which means a significant portion of our "Custom" code was able to move to purely "Generated."

### CODE REVIEW

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs [hunk 1]
```diff
@@ -15,28 +16,28 @@ namespace Azure.AI.OpenAI
...
         /// <inheritdoc cref="CompletionsOptions.ChoicesPerPrompt"/>
-        [CodeGenMember("N")]
         public int? ChoicesPerPrompt { get; set; }
 
         /// <inheritdoc cref="CompletionsOptions.FrequencyPenalty"/>
         public float? FrequencyPenalty { get; set; }
 
         /// <inheritdoc cref="CompletionsOptions.TokenSelectionBiases"/>
-        public IDictionary<int, int> TokenSelectionBiases { get; }
+        public IDictionary<int, int> TokenSelectionBiases
```
[the remark this answers is not in the document]
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs:25]: This deserves its own dedicated discussion -- tokens are honestly a mess!

Conceptually, this is a map from positive integer token IDs (as computed via a "black box" algorithm that's purely SDK-external for the time being) to [-100, 100] integer bias values.

The major reason I don't want to skip deserializing these keys into a proper type is that string keys end up inherently confusable with the source string tokens themselves.

As a grounded example: a developer is likely to approach use of this with a purpose in mind like "I want to make the word 'rain' less likely to appear in my completions responses about weather forecasts."
- The correct path for this today is to use a library like `tiktoken` or ahead-of-time conversion via something like https://platform.openai.com/tokenizer to discover that "rain" has an ID of 3201, and then assign `TokenSelectionBiases[3201]` to a value like `-50`
- But without a fair degree of *a priori* knowledge, a developer seeing an `IDictionary<string, int>` is very likely to try setting `TokenSelectionBiases["rain"]` to `-50`, instead; if we're lucky, that'll at least give a descriptive error message -- but it'd be better to not rely on luck and guide towards the right path ahead of time

We're looking into the possibility of having better SDK representation of the token concept, including potentially having client-side calculation. Then this might become something like an `IDictionary<OpenAIStringToken, int>`, which is the kind of clarity we'd love to have. But in the interim, the underlying data type for the key is definitely an `int` and we want it to show up that way.

Looked at from another angle, the *only* reason we'd have this as a string is because it happens to be serialized on the wire that way--and that's just a coincidence of convenience! E.g. if `"logit_bias": { "3201": -50 }` were instead a slightly more verbose `"logit_bias": [ { "key": 3201, "value": -50 } ]`, we'd never consider `string` as the user-facing type.

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs [hunk 2]
```diff
@@ -15,28 +16,28 @@ namespace Azure.AI.OpenAI
...
+        {
+            get
```
[the remark this answers is not in the document]
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs:27]: Per https://github.com/Azure/azure-sdk-for-net/issues/35096, we have a convention that at least *some* (List) collections in input models are (root, container) read-only, with contents written into them. That is, `Messages = new List<ChatMessage>() {...}` would be inconsistent (and also nonstandard for emission purposes).

If it's not unduly burdensome, I'd like to have consistency with collections in the input model since it seems confusing to not allow `Messages = new ...` but then allow `TokenSelectionBiases = new ...` in the same input model object.

Dictionary initializer syntax still works nicely, too, so e.g. this is possible:
```
var requestOptions = new CompletionsOptions()
{
    Prompts = { "What's the weather like in Seattle?" },
    TokenSelectionBiases =
    {
        [3201] = -50, // don't say "rain" as much
    },
};
```

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs [hunk 3]
```diff
@@ -15,28 +16,28 @@ namespace Azure.AI.OpenAI
     public partial class ChatCompletionsOptions
     {
         /// <inheritdoc cref="CompletionsOptions.ChoicesPerPrompt"/>
-        [CodeGenMember("N")]
         public int? ChoicesPerPrompt { get; set; }
```
[the remark this answers is not in the document]
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs:19]: Great question.

There's actually one level beyond that, even: unlike `/completions`, `/chat/completions` doesn't inherently "batch" -- it's effectively just one prompt with `Messages` forming the typed structure around roles, meaning there's not really a "per" involved at all.

If we're OK with renaming it (I can see arguments both ways), this could *really* be named something like `ChoicesCount`, `ChoicesToRequest`, or just `Choices`.

`ChoicesPerPrompt` isn't "wrong," but the non-intuitive catch is that there's only ever one prompt with Chat API structure!

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatMessage.cs [hunk 4]
```diff
@@ -10,27 +10,20 @@ namespace Azure.AI.OpenAI
     /// <summary> A single, role-attributed message within a chat completion interaction. </summary>
     public partial class ChatMessage
```
[the remark this answers is not in the document]
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatMessage.cs:11]: It should, but I get a `tsp compile` error if I don't define it here. Can follow up on that offline; as soon as *that's* resolved, there's no reason this custom class should exist.

File: sdk/openai/Azure.AI.OpenAI/src/Custom/Choice.Serialization.cs [hunk 5]
```diff
@@ -16,10 +16,10 @@ public partial class Choice
     {
         internal static Choice DeserializeChoice(JsonElement element)
```
Comment (joseharriaga) [inline sdk/openai/Azure.AI.OpenAI/src/Custom/Choice.Serialization.cs:17]: Can all this custom implementation be deleted and let it be auto-generated?

File: sdk/openai/Azure.AI.OpenAI/src/Custom/EmbeddingsOptions.cs [hunk 6]
```diff
@@ -13,6 +13,8 @@ namespace Azure.AI.OpenAI
     /// <summary> Schema to create a prompt completion from a deployment. </summary>
     public partial class EmbeddingsOptions
     {
+        internal string InternalNonAzureModelName { get; set; }
+
         /// <summary> Initializes a new instance of EmbeddingsOptions. </summary>
```
Comment (joseharriaga) [inline sdk/openai/Azure.AI.OpenAI/src/Custom/EmbeddingsOptions.cs:18]: I believe this constructor can be deleted from here and auto-generated.

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (m-nash) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs:26]: The generator doesn't support this today as there is no way to represent this in typespec we always assume Record<T> is a Dictionary<string, T>.

@johanste are there thoughts on whether typespec should allow you to specify the type of your key with some constraints on what would be valid here?

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs:28]: Oh, I apologize for missing the real problem here so blatantly! Alas, you're 100% right and my attempt to be clever and avoid the `Record<>` custom deserialization problem overlooked the most obvious issue -- namely that writing to the collection would actually end up in the request.

I went back to full custom serialization for the options classes (which gets rid of the awkward auto-conversion) and also added a bit of test coverage to ensure that the field is actually exercised.

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs:20]: I'll do that--if there's an issue with having different naming between the two, then the difference in actual behavior will be a good conversation point to have ;)

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatMessage.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatMessage.cs:11]: Odd, I see it working now, too -- either something else addressed that problem or the upgrade from 0.41 to 0.42 did.

I think I may still need to keep the class for now, though--we don't want "content" to be optional for user input, though it ends up (uniquely) optional when it's part of a streaming output.

The alternative would be that we make a parallel ChatMessage type just for the streaming response, but that seems even clunkier.

File: sdk/openai/Azure.AI.OpenAI/src/Custom/OpenAIClient.cs [hunk 7]
```diff
@@ -17,6 +17,8 @@ namespace Azure.AI.OpenAI
     /// <summary> Azure OpenAI APIs for completions and search. </summary>
     [CodeGenSuppress("GetCompletions", typeof(string), typeof(RequestContent), typeof(RequestContext))]
     [CodeGenSuppress("GetCompletionsAsync", typeof(string), typeof(RequestContent), typeof(RequestContext))]
+    [CodeGenSuppress("GetChatCompletions", typeof(string), typeof(RequestContent), typeof(RequestContext))]
+    [CodeGenSuppress("GetChatCompletionsAsync", typeof(string), typeof(RequestContent), typeof(RequestContext))]
     [CodeGenSuppress("GetEmbeddings", typeof(string), typeof(RequestContent), typeof(RequestContext))]
     [CodeGenSuppress("GetEmbeddingsAsync", typeof(string), typeof(RequestContent), typeof(RequestContext))]
     [CodeGenSuppress("CreateGetCompletionsRequest", typeof(string), typeof(RequestContent), typeof(RequestContext))]
```
[the remark this answers is not in the document]
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/OpenAIClient.cs:24]: Indeed it does! Great catch.

File: sdk/openai/Azure.AI.OpenAI/src/Generated/CompletionsLogProbabilityModel.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        /// <summary> The textual forms of tokens evaluated in this probability model. </summary>
+        public IReadOnlyList<string> Tokens { get; }
+        /// <summary> A collection of log probability values for the tokens in this completions data. </summary>
+        public IReadOnlyList<float> TokenLogProbabilities { get; }
+        /// <summary> A mapping of tokens to maximum log probability values in this completions data. </summary>
+        public IReadOnlyList<IDictionary<string, float>> TopLogProbabilities { get; }
+        /// <summary> The text offsets associated with tokens in this completions data. </summary>
+        public IReadOnlyList<int> TextOffsets { get; }
```
[the remark this answers is not in the document]
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Generated/CompletionsLogProbabilityModel.cs:46]: I believe it's just telling you the substring indices of what's in `tokens`. I don't see any particular utility in it given you can just look at the lengths in `tokens`.

E.g. if I have a partial `tokens` array of
```
"tokens": [
  "Are",
  " bananas",
  " especially",
  " radioactive",
```

I end up with `TextOffsets` (partial) of
```
"text_offset": [0, 3, 11, 22, 34],
```

Think we should just remove this from the projection?

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs [hunk 9]
```diff
@@ -53,38 +55,32 @@ public partial class ChatCompletionsOptions
...
         /// <summary> Initializes a new instance of ChatCompletionsOptions. </summary>
-        public ChatCompletionsOptions()
+        /// <param name="messages">
+        /// The collection of context messages associated with this chat completions request.
+        /// Typical usage begins with a chat message for the System role that provides instructions for
+        /// the behavior of the assistant, followed by alternating messages between the User and
+        /// Assistant roles.
+        /// </param>
+        /// <exception cref="ArgumentNullException"> <paramref name="messages"/> is null. </exception>
+        public ChatCompletionsOptions(IEnumerable<ChatMessage> messages)
```
[the remark this answers is not in the document]
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs:72]: I think this may boil down to a style/guideline question: do we want to enforce `messages` as a required constructor parameter on the options?

On one hand: it *is* required, and you'll get an error if you don't provide it; having it in the constructor provides compile-time enforcement of that.

On the other hand: it "feels weird" to have it stand out that way, and the code, as far as I can tell, doesn't flow as nicely.

E.g. right now, with the constructor parameter optional, I have:
```
            var requestOptions = new ChatCompletionsOptions()
            {
                Messages =
                {
                    new ChatMessage(ChatRole.System, "You are a helpful assistant."),
                    new ChatMessage(ChatRole.User, "Can you help me?"),
                    new ChatMessage(ChatRole.Assistant, "Of course! What do you need help with?"),
                    new ChatMessage(ChatRole.User, "What temperature should I bake pizza at?"),
                },
                MaxTokens = 512,
            };
```

With it required, that'd change to:
```
            var requestOptions = new ChatCompletionsOptions(new ChatMessage[]
            {
                    new ChatMessage(ChatRole.System, "You are a helpful assistant."),
                    new ChatMessage(ChatRole.User, "Can you help me?"),
                    new ChatMessage(ChatRole.Assistant, "Of course! What do you need help with?"),
                    new ChatMessage(ChatRole.User, "What temperature should I bake pizza at?"),
            })
            {
                MaxTokens = 512,
            };
```

Which requires choosing a collection type for initialization and a much less straightforward brace-initialization pattern. It also adds an extra logical deviation from the REST structure for dubious utility, though that's not a good justification on its own without more.

Is that awkwardness just in my head and part of our existing guidelines?

File: sdk/openai/Azure.AI.OpenAI/src/Custom/Choice.Serialization.cs (hunk 5 above)
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/Choice.Serialization.cs:17]: Unfortunately, no--though I should (and will now) add a comment into the custom code to clarify why.

Right now, the generated version yields this for some requests:
```
  Message: 
System.Text.Json.JsonException : A property 'logprobs' defined as non-nullable but received as null from the service. This exception only happens in DEBUG builds of the library and would be ignored in the release build

  Stack Trace: 
JsonElementExtensions.ThrowNonNullablePropertyIsNull(JsonProperty property) line 105
Choice.DeserializeChoice(JsonElement element) line 42
```

This may be a generalized version of the "nullable float" problem that was addressed, though in this case making `logprobs?` in `Choice` `CompletionsLogProbabilityModel | null` doesn't seem to address it.

File: sdk/openai/Azure.AI.OpenAI/src/Custom/EmbeddingsOptions.cs (hunk 6 above)
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/EmbeddingsOptions.cs:18]: Awesome, indeed it can!

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (joseharriaga) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs]: This is merely a thought, but potentially we can bring the naming closer together if we additionally rename the property in `CompletionsOptions` something like `ChoiceCountPerPrompt`. This makes it clear that these two properties are analogous but with a subtle difference in behavior. We can think more about it during the architecture review. 😊

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (joseharriaga) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs:61]: Looking at the [.NET Guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html), we have the following:
> ✅ DO provide a minimal constructor that takes only the parameters required to connect to the service.

But this is technically referring to the client constructors, and here we're talking about a model class. Though I think the spirit of the guideline still applies here: Use the constructor to ensure that you have a "minimum viable" instance. 

In your example above, is your concern partially because you think that using `MaxTokens` might be particularly common? If so, I wonder if the "right" approach is to let the emitter generate the constructors with the required parameters only, and then add custom "convenience" constructors that additionally have common optional parameters.

File: sdk/openai/Azure.AI.OpenAI/src/Custom/Choice.Serialization.cs (hunk 5 above)
Comment (joseharriaga) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/Choice.Serialization.cs:17]: Looking at the cadl.json that is produced during code generation, I see the `LogProbabilityModel` property of the `Choice` model and that its `IsNullable` value is `true`. Therefore, I think this is probably a bug in the emitter that is causing it to not respect this value.

@m-nash: What do you think? I'm thinking we keep the custom serialization for now, and I'll file an issue for this in the [autorest.csharp repo](https://github.com/Azure/autorest.csharp). Do you know if it needs any special tags or anything in particular?

File: sdk/openai/Azure.AI.OpenAI/src/Generated/CompletionsLogProbabilityModel.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (joseharriaga) [reply sdk/openai/Azure.AI.OpenAI/src/Generated/CompletionsLogProbabilityModel.cs:46]: Gotcha. Thank you for the explanation! This doesn't need to be considered as part of this PR, but it got me thinking because I've seen something similar in other Azure services like Text Analytics and Form Recognizer. The interesting thing about these offsets is that the text encoding that is used can affect them and the expectations on how to use them, especially when using emoji and multilingual text. For context, .NET uses UTF-16 by default to encode the text in a `string`, so when using methods like `Substring()`, the expectation is that the offsets and lengths are adjusted for UTF-16.

Here's more info about the problem and the approach taken by the Azure Cognitive Service for Language: https://learn.microsoft.com/en-us/azure/cognitive-services/language-service/concepts/multilingual-emoji-support

This might not be particularly important in this case, since, like you say, the utility of these `TextOffsets` is questionable.

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatMessage.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (joseharriaga) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatMessage.cs:11]: That makes sense. Let's keep it.

**Review by joseharriaga — APPROVED** [review]

File: sdk/openai/Azure.AI.OpenAI/src/Custom/CompletionsFinishReason.cs [hunk 10]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (joseharriaga) [inline sdk/openai/Azure.AI.OpenAI/src/Custom/CompletionsFinishReason.cs:1]: This custom implementation can now be fully removed.

File: sdk/openai/Azure.AI.OpenAI/src/Custom/Choice.Serialization.cs (hunk 5 above)
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/Choice.Serialization.cs]: Thanks, @m-nash. Updated to latest and confirmed that it works!

File: sdk/openai/Azure.AI.OpenAI/src/Custom/StreamingChatChoice.cs [hunk 11]
```diff
@@ -33,7 +33,8 @@ public class StreamingChatChoice
         /// Normal termination typically provides "stop" and encountering token limits in a request typically
         /// provides "length." If no value is present, this StreamingChoice is still in progress.
         /// </remarks>
-        public string FinishReason => GetLocked(() => _baseChoices.Last().FinishReason);
+        public CompletionsFinishReason? FinishReason => GetLocked(() =>
+            _baseChoices.Last().FinishReason == null ? null : _baseChoices.Last().FinishReason);
```
[the remark this answers is not in the document]
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/StreamingChatChoice.cs:37]: Versus using the `GetLocked` synchronization helper? I wasn't sure what would end up being concurrency-safe via LINQ since `_baseChoices` can be updated while it's being read here, though intuitively I would hope `.Last()` isn't problematic for a purely additive case.

Do you see any issues with keeping all the concurrent collection use under the lock just to keep it consistent, though?

File: sdk/openai/Azure.AI.OpenAI/src/Custom/Choice.Serialization.cs [hunk 12]
```diff
@@ -16,10 +16,10 @@ public partial class Choice
...
         {
-            Optional<string> text = default;
```
[the remark this answers is not in the document]
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/Choice.Serialization.cs:19]: Agreed on minimizing discrepancies! Fortunately, with the latest update, this custom file was able to be removed entirely.

File: sdk/openai/Azure.AI.OpenAI/src/Custom/CompletionsFinishReason.cs (hunk 10 above)
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/CompletionsFinishReason.cs:1]: @joseharriaga I unfortunately still couldn't get it to work; the camel-to-Pascal casing issue is resolved, but the renames (likely no current use of `projectedName("json", ...)`) don't "stick" -- e.g. this typespec definition:

```
    @doc("Completions ended normally and reached its end of token generation.")
    @projectedName("json", "stop")
    stopped,
```

...still uses the underlying string "stopped" instead of the wire format of "stop."

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs [hunk 13]
```diff
@@ -50,41 +38,39 @@ public partial class ChatCompletionsOptions
...
         public float? Temperature { get; set; }
 
+        /// <inheritdoc cref="CompletionsOptions.TokenSelectionBiases"/>
+        public IDictionary<int, int> TokenSelectionBiases { get; }
+
         /// <inheritdoc cref="CompletionsOptions.User"/>
         public string User { get; set; }
 
-        internal string NonAzureModel { get; set; }
+        internal IDictionary<string, int> InternalStringKeyedTokenSelectionBiases { get; }
```
[the remark this answers is not in the document]
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs:47]: Good catch! Missed removing this when updating the custom serialization logic. I'll remove those now.

[the remark this answers is not in the document]
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs:47]: Crud, no, I spoke too soon.

Right now, as long as we have `logit_bias` defined in the typespec as a `Record<int32>`, it'll demand a generated `Dictionary<string, int>` *somewhere*. To make it work, it's given a clearly funky projected name (with "internal" at the beginning) and then marked with `internal` visibility to hide it (in favor of the custom `Dictionary<int, int>`).

Is there a better way to achieve this?

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs:61]: Setting `MaxTokens`, `Temperature`, and a few other settings is likely to be extremely common even if not outright required, yes--and that is part of the concern around the awkwardness of mixed constructor/init-setter use per the earlier.

More than that, though, is the contrast to the established "flatness" in the REST and Python usage, e.g. in OpenAI's quickstart (python/Node.js) here: https://platform.openai.com/docs/quickstart/understand-the-code -- there, there's no structural delineation made between required and optional fields; it's just a big, flat property bag.

The advantage I see of promoting functionally required request payload members to the options constructor is the forced clarity of making viable request payloads; here, I think we don't gain much from that and that the cognitive burden of having to define common request fields (some "required," some not) in two different places outweighs the gain.

A somewhat more contemporary compromise might be init-only setters, though I'm not sure where guidelines land on that (they call out get/set semantics but I don't see any mention of init).

Would you object to keeping things "flat" for the time being and noting this pattern as a discussion topic for the next round of holistic surface review?

File: sdk/openai/Azure.AI.OpenAI/src/Custom/CompletionsOptions.cs
Comment (joseharriaga) [inline sdk/openai/Azure.AI.OpenAI/src/Custom/CompletionsOptions.cs:1]: In this PR, I see that we removed several uses of the `[CodeGenMember]` attribute in favor of doing the renames via the `@projectedName` decorator in TypeSpec. Doing these customizations in TypeSpec is great and definitely the way to go! I see that we still have a few remaining uses of the `[CodeGenMember]` attribute though. Is it possible to remove those too in the same manner?

Comment (trrwilson) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/CompletionsOptions.cs:1]: Absolutely! Those can be entirely removed now. `@projectedName` is a far, far nicer way to achieve it. All instances deleted in the next commit.

File: sdk/openai/Azure.AI.OpenAI/api/Azure.AI.OpenAI.netstandard2.0.cs [hunk 14]
```diff
@@ -3,22 +3,23 @@ namespace Azure.AI.OpenAI
     public partial class ChatChoice
     {
         internal ChatChoice() { }
-        public string FinishReason { get { throw null; } }
-        public int? Index { get { throw null; } }
+        public Azure.AI.OpenAI.CompletionsFinishReason FinishReason { get { throw null; } }
```
Comment (joseharriaga) [inline sdk/openai/Azure.AI.OpenAI/api/Azure.AI.OpenAI.netstandard2.0.cs:6]: We talked offline about customizing the constructors of the other classes that have nullable properties to manually remove the `Argument.AssertNotNull` check as a workaround for the following issue:
🔗 https://github.com/Azure/autorest.csharp/issues/2340

I'm noticing that this one here is a little bit of a special case because it involves a nullable extensible enum property.

1. Here, the `FinishReason` property is not defined as nullable (i.e., `CompletionsFinishReason? FinishReason`). This is important in this case because `CompletionsFinishReason` is a struct (i.e., a value type).

2. The generated `DeserializeChatCompletions` method deserializes the `FinishReason` property like this:
```csharp
if (property.NameEquals("finish_reason"u8)) 
{
    finishReason = new CompletionsFinishReason(property.Value.GetString());
    continue;
}
```
When the service returns null, I believe this would throw a `NullReferenceException`? But even if it didn't, this code would new up a `CompletionsFinishReason` object, and while its internal value would be null, what we really want is for there to be no object at all; `FinishReason` should be set to null.

In other words, the workaround in this case requires customizing this property and customizing this class's serialization code. Having a test for this would also be useful; I'm thinking maybe something simple like mocking a response from the service where this is null and verifying that everything works as expected.

I have added all this information to the existing issue in the autorest.csharp repo that I linked above.

File: sdk/openai/Azure.AI.OpenAI/src/Custom/StreamingChatChoice.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (joseharriaga) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/StreamingChatChoice.cs:37]: No, I was just referring to reducing this:
```csharp
_baseChoices.Last().FinishReason == null ? null : _baseChoices.Last().FinishReason
```

To this:
```
_baseChoices.Last().FinishReason
```

Like it is in StreamingChoice.cs. In other words, you don't need to check for null.

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (joseharriaga) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs:60]: Sounds good. 👍

File: sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (joseharriaga) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/ChatCompletionsOptions.cs:46]: You can add `[CodeGenSuppress("InternalStringKeyedTokenSelectionBiases")]` as an attribute on this class. I feel that between adding this attribute or adding a property to make it internal and never use it, the attribute might be the cleaner choice.

File: sdk/openai/Azure.AI.OpenAI/src/Custom/CompletionsFinishReason.cs (hunk 10 above)
Comment (joseharriaga) [reply sdk/openai/Azure.AI.OpenAI/src/Custom/CompletionsFinishReason.cs:1]: Got it. It's unfortunate that `@projectedName` doesn't work on enum values at the moment. 😢 

Then, here's what I'm thinking:
1. Since this is a customization that we want in C# only (and possibly Java?), the names of these values in the TypeSpec should match the REST API for consistency instead. We're not winning much by renaming them in the TypeSpec in any case, so that makes a stronger argument in favor of that change.
2. We can then change this custom class to suppress the members with the wrong names, while also removing the custom constructor. In other words, we would end up with the following:

```csharp
/// <summary> Representation of the manner in which a completions response concluded. </summary>
[CodeGenSuppress("Stop")]
[CodeGenSuppress("StopValue")]
[CodeGenSuppress("Length")]
[CodeGenSuppress("LengthValue")]
[CodeGenSuppress("ContentFilter")]
[CodeGenSuppress("ContentFilterValue")]
public readonly partial struct CompletionsFinishReason
{
    private const string StoppedValue = "stop";
    private const string TokenLimitReachedValue = "length";
    private const string ContentFilteredValue = "content_filter";

    /// <summary> Completions ended normally and reached its end of token generation. </summary>
    public static CompletionsFinishReason Stopped { get; } = new CompletionsFinishReason(StoppedValue);
    /// <summary> Completions exhausted available token limits before generation could complete. </summary>
    public static CompletionsFinishReason TokenLimitReached { get; } = new CompletionsFinishReason(TokenLimitReachedValue);
    /// <summary>
    /// Completions generated a response that was identified as potentially sensitive per content
    /// moderation policies.
    /// </summary>
    public static CompletionsFinishReason ContentFiltered { get; } = new CompletionsFinishReason(ContentFilteredValue);
}
```

I feel like this is relatively clean for a customization? What do you think?
